File: 2017-kotlin/day01.kt
import java.io.File

fun step_sum(t: String, d: Int): Int {
    var sum = 0
    for ((i, x) in t.withIndex()) {
        if (x == t[(i+d) % t.length]) {
            sum += x.digitToInt()
        }
    }
    return sum
}

fun main() {
    val t = File("/Users/xdavidliu/Documents/aoc/input01.txt")
        .readText().trimEnd()
    val p1 = step_sum(t, 1)
    println("part 1 = $p1")  // 1029
    val p2 = step_sum(t, t.length / 2)
    println("part 2 = $p2")  // 1220
}

File: 2024/day01.swift
func getTwoArrays(lines: [String]) -> ([Int], [Int]) {
    var a: [Int] = [], b: [Int] = []
    for line in lines {
        let toks = line.split(separator: " ")
        a.append(Int(toks[0])!)
        b.append(Int(toks[1])!)
    }
    a.sort()
    b.sort()
    return (a, b)
}

func getCounter(nums: [Int]) -> [Int: Int] {
    var d: [Int: Int] = [:]
    for x in nums {
        d[x, default: 0] += 1
    }
    return d
}

// hack: if put in Documents folder, it will keep asking permission.
// astonishingly, home dir does not have this security
let lines = getLines(fileName: "/Users/xdavidliu/input01.txt")
let (a, b) = getTwoArrays(lines: lines)
var sum = 0
for (x, y) in zip(a, b) {
    sum += abs(x - y)
}
print("part 1 = \(sum)")
// 1970720

var p2 = 0
let bCount = getCounter(nums: b)
for x in a {
    p2 += x * bCount[x, default: 0]
}
print("part 2 = \(p2)")
// 17191599
